### DISCUSSION TITLE
[help needed] clarification for master client and volume growth

### DISCUSSION THREAD

**Discussion by andypeng2015** [seaweedfs/seaweedfs#2468] [Q&A] [answered]
Hi I have some doubts after reading the code, hope to get some clarification here

1: what's the use case / purpose of having master client to connect to all masters? since we already have raft to connect all masters, raft is only used for leader election?

https://github.com/chrislusf/seaweedfs/blob/master/weed/wdclient/masterclient.go#L49

https://github.com/chrislusf/seaweedfs/blob/master/weed/wdclient/masterclient.go#L97

2: how the volume grow works? Does it create multiple logic volume on data nodes to increase concurrency only? In the wiki it says "One volume serves one write a time. If you need to increase concurrency, you can pre-allocate lots of volumes.". 

https://github.com/chrislusf/seaweedfs/blob/master/weed/topology/volume_growth.go#L90

thanks in advance

**Comment by chrislusf** [comment, marked as answer]
1. raft is only for leader election.
2. "Does it create multiple logic volume on data nodes to increase concurrency only?" Yes.

**Reply by andypeng2015** [reply]
@chrislusf thanks for the prompt reply, just to confirm, master client from other masters will connect to master leader only (single source of truth) to exchange volume info?

**Reply by chrislusf** [reply]
correct

**Reply by andypeng2015** [reply]
Hi @chrislusf I have one last clarification for this topic.

Since Raft is only used for leader election and other masters pull volume info from leader, it's eventual consistency.

Why we donot use gossip protocol to exchange volume info among the masters, and if it converge fast enough (since only 3 to 5 masters), it's also eventual consistency. Any particular reason here to opt in Raft?

**Reply by chrislusf** [reply]
what problem this change can address?

**Reply by andypeng2015** [reply]
Not for a problem, have not encounter any problem with masters.

Just to confirm the rational behind the design, IMO the gossip protocol without leader election should also work, as I described above, just wondering why it uses Raft.

**Reply by chrislusf** [reply]
to my experience, gossip protocol is not fast enough and can often lead to inconsistent data.
